test/request_reply_message_receiver_test.go

```diff
@@ -34,81 +39,801 @@ const subscriptionString = "some-subscription"
 
-		helpers.ConnectMessagingService(messagingService)
-		helpers.DisconnectMessagingService(messagingService)
+				expected := &solace.IllegalArgumentError{}
+				Expect(err).ToNot(BeAssignableToTypeOf(expected), fmt.Sprintf("Expected error of type %T to not be assignable of type %T", err, expected))
+			})
+		}
 
-		err = receiver.Start()
-		Expect(err).To(HaveOccurred())
-		Expect(err).To(BeAssignableToTypeOf(&solace.IllegalStateError{}))
 	})
 
-	It("fails to receive a message on an unconnected receiver", func() {
-		receiver, err := messagingService.RequestReply().CreateRequestReplyMessageReceiverBuilder().Build(subscription)
-		Expect(err).ToNot(HaveOccurred())
+	Context("with a messaging service that will be disconnected", func() {
+		const topicString = "terminate/me"
+		const messagesPublished = 3
+		const publishTimeOut = 3 * time.Second
+		var terminationChannel chan solace.TerminationEvent
+		var receiver solace.RequestReplyMessageReceiver
+
+		BeforeEach(func() {
+			helpers.ConnectMessagingService(messagingService)
+			var err error
+			receiver, err = messagingService.RequestReply().CreateRequestReplyMessageReceiverBuilder().Build(resource.TopicSubscriptionOf(topicString))
+			Expect(err).ToNot(HaveOccurred())
+			err = receiver.Start()
+			Expect(err).ToNot(HaveOccurred())
+
+			terminationChannel = make(chan solace.TerminationEvent)
+			startTime := time.Now()
+			receiver.SetTerminationNotificationListener(func(te solace.TerminationEvent) {
+				defer GinkgoRecover()
+				Expect(te.GetTimestamp()).To(BeTemporally(">=", startTime))
+				Expect(te.GetTimestamp()).To(BeTemporally("<=", time.Now()))
+				Expect(te.GetCause()).ToNot(BeNil())
+				Expect(te.GetMessage()).ToNot(BeEmpty())
+				terminationChannel <- te
+			})
+		})
+
+		AfterEach(func() {
+			if !receiver.IsTerminated() {
+				receiver.Terminate(1 * time.Second)
+			}
+			if messagingService.IsConnected() {
+				messagingService.Disconnect()
+			}
+		})
+
+		forceDisconnectFunctions := map[string]func(messagingService solace.MessagingService){
+			"messaging service disconnect": func(messagingService solace.MessagingService) {
+				helpers.DisconnectMessagingService(messagingService)
+			},
+			"SEMPv2 disconnect": func(messagingService solace.MessagingService) {
+				helpers.ForceDisconnectViaSEMPv2(messagingService)
+			},
+		}
+
+		for testCase, disconnectFunctionRef := range forceDisconnectFunctions {
+			disconnectFunction := disconnectFunctionRef
+			It("terminates the receiver using async receive when disconnecting with "+testCase, func() {
+				blocker := make(chan struct{})
+				msgsReceived := make(chan message.InboundMessage, messagesPublished)
+
+				receiver.ReceiveAsync(func(inboundMessage message.InboundMessage, replier solace.Replier) {
+					<-blocker
+					msgsReceived <- inboundMessage
+
+					payload, _ := inboundMessage.GetPayloadAsString()
+					if replier != nil {
+						replier.Reply(helpers.NewMessage(messagingService, "Reply for: "+payload))
```

Comment:
Is this missing an Expect statement on the error return from Reply.
Is there a test to verify a call to Replier.reply returns an error if the receiver is terminated or service is disconnected. As a part the lifecycle behaviour tests?